src/common/ad_mem.c

```diff
@@ -261,18 +292,84 @@ blob_fini(struct ad_blob *blob)
 	while ((arena = d_list_pop_entry(&blob->bb_ars_lru, struct ad_arena, ar_link)))
 		free_arena(arena, true);
 
+	d_binheap_destroy(&blob->bb_arena_free_heap);
+	D_FREE(blob->bb_mh_nodes);
 	D_FREE(blob->bb_bmap_rsv);
 	D_FREE(blob->bb_pages);
 	D_FREE(blob->bb_mmap);
 	return 0;
 }
 
+static uint32_t
+arena_free_size(struct ad_arena_df *ad)
+{
+	int		    grp_idx;
+	uint32_t	    weight = 0;
+	struct ad_group_df  *gd;
+	uint32_t	    grp_max = (ARENA_SIZE + ARENA_UNIT_SIZE - 1) / ARENA_UNIT_SIZE;
+
+	for (grp_idx = 0; grp_idx < grp_max; grp_idx++) {
+		if (!isset64(ad->ad_bmap, grp_idx)) {
+			weight += (1 << GRP_SIZE_SHIFT);
+			continue;
+		}
+	}
+
+	for (grp_idx = 0; grp_idx < ad->ad_grp_nr; grp_idx++) {
+		gd = &ad->ad_groups[grp_idx];
+		weight += gd->gd_unit_free * gd->gd_unit;
+	}
+
+	return weight;
+}
+
+#define ARENA_WEIGHT_BITS 12
```

Comment:
if application always requests 4K extent, we have to reorder arenas on each alloc/free? I would choose 14 at least (16K, 0.1% of the arena size)